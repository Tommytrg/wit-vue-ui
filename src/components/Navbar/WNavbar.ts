export type NavLinks = Array<NavLink>
export type NavLink = { key: string; active: boolean; to: string; rel: string; locale: string }
