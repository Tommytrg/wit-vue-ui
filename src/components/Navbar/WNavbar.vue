<template>
  <header class="navbar">
    <div class="logo-link">
      <slot class="logo-link" name="logo"></slot>
    </div>
    <input id="side-menu" ref="checkbox" class="checkbox" type="checkbox" />
    <nav class="navbar-tabs">
      <ul class="navbar-links animated zoomIn">
        <div class="close-menu-container" @click="closeMenu">
          <CloseNavIcon />
        </div>
        <li v-for="link in navLinks" :key="link.key" class="nav-link" @click="closeMenu">
          <a
            :href="link.to"
            class="text font-[500] flex cursor-pointer sm:text-white-50 relative p-[8px] m-[-8px] pl-[16px]"
            @mouseover="activateLink(link)"
            @mouseleave="clearActiveLinks"
          >
            <Transition name="fade">
              <NavigationCursor
                v-if="link.active"
                class="w-sm h-auto absolute top-[16px] left-[4px]"
              />
            </Transition>
            <span /><span class="slash text-wit-blue-500">/</span>{{ link.locale }}
          </a>
        </li>
        <!-- <li class="nav-link language-select">
          <client-only>
            <LanguageSwitcher />
          </client-only>
        </li> -->
      </ul>
    </nav>
    <label ref="menu" for="side-menu" class="hamburguer-container">
      <HamburgerIcon class="hamburguer" />
    </label>
  </header>
</template>

<script lang="ts" setup>
import { ref, PropType, watch, computed } from 'vue'
import { NavLink, NavLinks } from './WNavbar.ts'
import NavigationCursor from '@/assets/svg/navigation-cursor.svg?component'
import HamburgerIcon from '@/assets/svg/hamburguer.svg?component'
import CloseNavIcon from '@/assets/svg/close-nav.svg?component'
const menu = ref()
const checkbox = ref()

const props = defineProps({
  navLinks: {
    type: Object as PropType<NavLinks>,
    required: true
  },
  redirectionActive: {
    type: Boolean,
    required: true
  }
})

const localNabLinks = ref(props.navLinks)

const redirectionActive = computed(() => props.redirectionActive)

function closeMenu() {
  menu.value.click()
}

function activateLink(link: NavLink) {
  link.active = true
}
function clearActiveLinks() {
  localNabLinks.value.forEach((link: NavLink) => (link.active = false))
}

watch(redirectionActive, () => {
  if ((checkbox.value as HTMLInputElement).checked) {
    closeMenu()
  }
})
</script>

<style lang="scss">
.navbar {
  background-color: transparent;
  display: flex;
  justify-content: space-between;
  align-items: center;
  top: -16px;
  z-index: 500;

  .logo-link {
    a {
      text-decoration: none;
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
  .close-menu-container {
    display: none;
  }
  .checkbox {
    display: none;
  }
  .navbar-links {
    display: grid;
    grid-template-columns: repeat(4, max-content);
    align-items: flex-start;
    justify-content: space-between;
    .nav-link {
      padding: 0 16px;
      text-decoration: none;
      font-size: 16px;
      text-decoration: none;
      width: max-content;
      display: block;
      &.language-select {
        padding: 16px;
        justify-self: flex-end;
        width: max-content;
        display: none;
      }
    }
  }
  .hamburguer-rect {
    @apply fill-black-950 dark:fill-white-50;
  }
  .hamburguer-container {
    display: none;
    .hamburguer {
      width: 30px;
      height: auto;
    }
  }
}
@keyframes zoomIn {
  from {
    opacity: 0;
    -webkit-transform: scale3d(0.3, 0.3, 0.3);
    transform: scale3d(0.3, 0.3, 0.3);
  }

  50% {
    opacity: 1;
  }
}

.zoomIn {
  -webkit-animation-name: zoomIn;
  animation-name: zoomIn;
}

.animated {
  -webkit-animation-duration: 0.3s;
  animation-duration: 0.3s;
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}
@media (max-width: 706px) {
  .navbar {
    .navbar-tabs {
      display: none;
      .navbar-links {
        height: 100vh;
      }
    }
    .checkbox:checked ~ .navbar-tabs {
      display: block;
    }
    .close-menu-container {
      display: none;
    }
    .hamburguer-container {
      display: flex;
      justify-content: center;
      align-items: center;
    }
    .navbar-links {
      @apply bg-black-950 dark:bg-black-900;
      padding: 0;
      display: grid;
      grid-template-rows: repeat(4, max-content);
      grid-template-columns: 1fr;
      position: fixed;
      width: 100vw;
      z-index: 500;
      right: 0;
      left: 0;
      bottom: 0;
      top: 0;
      gap: 0;
      .nav-link {
        padding: 8px 16px 8px 16px;
        width: 100%;
        &.language-select {
          display: block;
          padding: 0px;
          margin-right: 8px;
          justify-self: flex-end;
        }
      }
      .close-menu-container {
        padding: 16px;
        display: flex;
        justify-content: flex-end;
      }
      li {
        height: max-content;
      }
    }
  }
}
</style>
