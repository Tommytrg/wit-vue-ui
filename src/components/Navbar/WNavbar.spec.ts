import { describe, it, expect } from 'vitest'

import { mount } from '@vue/test-utils'
import WNavbar from './WNavbar.vue'

describe('WNavbar', () => {
  const wrapper = mount(WNavbar, {
    props: {
      navLinks: [
        {
          key: 'stake',
          active: false,
          to: '/stake',
          rel: 'stake',
          locale: 'Stake'
        },
        {
          key: 'buy',
          active: false,
          to: '/buy',
          rel: 'buy',
          locale: 'Buy'
        },
        {
          key: 'build',
          active: true,
          to: '/build',
          rel: 'build',
          locale: 'Build'
        }
      ],
      redirectionActive: false
    }
  })
  it('renders properly first link', () => {
    expect(wrapper.html()).toContain('<a href="#stake"')
  })
  it('renders properly second link', () => {
    expect(wrapper.html()).toContain('<a href="#buy"')
  })
  it('renders properly third link', () => {
    expect(wrapper.html()).toContain('<a href="#build"')
  })
  it('renders properly pointer icon when active', () => {
    expect(wrapper.html())
      .toContain(`<li class="nav-link"><a href="#build" class="text font-[500] flex cursor-pointer sm:text-white-50 relative p-[8px] m-[-8px] pl-[16px]">
          <transition-stub name="fade" appear="false" persisted="false" css="true"><svg xmlns="http://www.w3.org/2000/svg" width="12" height="13" fill="none" viewBox="0 0 12 13" class="w-sm h-auto absolute top-[16px] left-[4px]">
              <path fill="#00E2ED" d="M11.17 6.39 5.89 9.44.61 12.48V.29l5.28 3.05z"></path>
            </svg></transition-stub><span></span><span class="slash text-wit-blue-500">/</span>Build
        </a></li>`)
  })
})
