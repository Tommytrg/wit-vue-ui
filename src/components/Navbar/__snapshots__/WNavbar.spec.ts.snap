// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WNavbar > renders properly the component 1`] = `
<header
  class="navbar"
>
  <div
    class="logo-link"
  >
    
    
  </div>
  <input
    class="checkbox"
    id="side-menu"
    type="checkbox"
  />
  <nav
    class="navbar-tabs"
  >
    <ul
      class="navbar-links animated zoomIn"
    >
      <div
        class="close-menu-container"
      >
        <svg
          fill="#fff"
          height="32"
          viewBox="0 0 32 32"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.333 12.333 16 16m0 0 3.666 3.666M16 16l-3.667 3.666M16 16l3.666-3.667M16 27C9.925 27 5 22.075 5 16S9.925 5 16 5s11 4.925 11 11-4.925 11-11 11"
            stroke="#292D32"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
      
      <li
        class="nav-link"
      >
        <a
          class="text font-[500] flex cursor-pointer sm:text-white-50 relative p-[8px] m-[-8px] pl-[16px]"
          href="#stake"
        >
          <transition-stub
            appear="false"
            css="true"
            name="fade"
            persisted="false"
          >
            <!--v-if-->
          </transition-stub>
          <span />
          <span
            class="slash text-wit-blue-500"
          >
            /
          </span>
          Stake
        </a>
      </li>
      <li
        class="nav-link"
      >
        <a
          class="text font-[500] flex cursor-pointer sm:text-white-50 relative p-[8px] m-[-8px] pl-[16px]"
          href="#buy"
        >
          <transition-stub
            appear="false"
            css="true"
            name="fade"
            persisted="false"
          >
            <!--v-if-->
          </transition-stub>
          <span />
          <span
            class="slash text-wit-blue-500"
          >
            /
          </span>
          Buy
        </a>
      </li>
      <li
        class="nav-link"
      >
        <a
          class="text font-[500] flex cursor-pointer sm:text-white-50 relative p-[8px] m-[-8px] pl-[16px]"
          href="#build"
        >
          <transition-stub
            appear="false"
            css="true"
            name="fade"
            persisted="false"
          >
            <svg
              class="w-sm h-auto absolute top-[16px] left-[4px]"
              fill="none"
              height="13"
              viewBox="0 0 12 13"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.17 6.39 5.89 9.44.61 12.48V.29l5.28 3.05z"
                fill="#00E2ED"
              />
            </svg>
          </transition-stub>
          <span />
          <span
            class="slash text-wit-blue-500"
          >
            /
          </span>
          Build
        </a>
      </li>
      
      <!-- &lt;li class="nav-link language-select"&gt;
          &lt;client-only&gt;
            &lt;LanguageSwitcher /&gt;
          &lt;/client-only&gt;
        &lt;/li&gt; -->
    </ul>
  </nav>
  <label
    class="hamburguer-container"
    for="side-menu"
  >
    <svg
      class="hamburguer"
      fill="none"
      height="37"
      viewBox="0 0 57 37"
      width="57"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M.821.4h56v8h-56zM.821 28.4h56v8h-56zM.821 14.4h56v8h-56z"
        fill="#232323"
      />
    </svg>
  </label>
</header>
`;
